src/main/java/org/lisoft/lsml/model/item/AmmoWeapon.java

```diff
@@ -45,25 +45,36 @@ public class AmmoWeapon extends Weapon {
     @XStreamAsAttribute
     private final String ammoTypeId;
     @XStreamAsAttribute
-    private final boolean oneShot;
-
+    private final int volleySize;
+    @XStreamAsAttribute
+    private final int ammoPerShot;
+    @XStreamAsAttribute
+    private final boolean oneShot;    
+    @XStreamAsAttribute
+    private final double firingDelay;
```

Comment:
Please don't store the derived value in the class, prefer to calculate it when needed (unless it's notably expensive to compute).

A unit test for a method call can be viewed as a black box with `x + y` bits of input and `z + y` bits of output. Where `x` is the number of total bits in the arguments to the method, `y` is the number of bits stored in the class and `z` is the number of bits in the returned output.

Therefore the number of possible input combinations to the method under test is `2^(x+y)`, meaning that for an exhaustive test suite, that's how many tests you need to write. And massaging the `y` bits into the correct state is very tedious and tends to break good class design. In practice we redue the number of tests by chosing a representative subset of the exhaustive tests focusing on covering all branches, corner cases, happy path and edges of the inputs. 

However, by increasing the input size to the test, the number of tests that are necessary invariably increases. Please try to always reduce the number of variables (the state) in the class.

Reply:
Well that is counter intuitive.  For something that will never change it seems like it should just be calculated one time and be done with it.  I had already added some testing.

What I noticed is that the LSML handler is asking for this value 8 to 10 times for each row of the weapons table.  I cant image how many times this value will be requested on the mech build page.  But I suspect four calls per loaded ballistic or missile weapon.  No it's not a huge calc... but it gets called a lot.

I can put it back into the method.

Reply:
This is another case of premature optimization, but at the cost of testability this time.

It's like 20 instructions to compute this, and on a 10 year old CPU with 10 IPC and 2 GHz, the time spent is around `(20 instructions /10 instructions per clock)/(2E9 clocks per second) = 1ns`. So even if it was done 1000 times per row of the weapons table, it'd still only take a micro second for each row. And the row is only updated when something changes on the mech that affects this attribute.

That said, the JIT optimizer is pretty good and it will do it's magic during runtime, based on what contributes the most to performance. It'll do inlining, constant propagation and common sub-expression elimination etc to avoid re-computing things. Please trust the compiler to do it's job, modern compilers are REALLY good and more often than not, the compiler will produce the same exact machine code for the hand optimized code and the readable code, so always perfer to write the readable/testable code by default.

If there is evidence that the program is too slow or that the JIT optimizer isn't doing a good enough job, then please profile firest and base any optimizations on those profile results.

From my 20 or so years in the industry writing performance critical code, performance is gained by having the right algorithm and through system level design, not by optimizing sub routines. Only in rare circumstances is hand optimizing code worthwhile, and it should only ever be done under the guidance of benchmarking and profiling results. Because otherwise you may spend hours working on something that won't actually affect the runtime of the application, or worst case, make it slower.

Reply:
ok.  I'll move it into the method.

Reply:
done